src/main/java/org/apache/pulsar/io/jcloud/partitioner/AbstractPartitioner.java

```diff
@@ -63,4 +72,27 @@ public String generatePartitionedPath(String topic, String encodedPartition) {
         joinList.add(encodedPartition);
         return StringUtils.join(joinList, PATH_SEPARATOR);
     }
+
+    protected long getMessageOffset(Record<T> record) {
+        if (useIndexAsOffset && record.getMessage().isPresent()) {
+            final Message<T> message = record.getMessage().get();
+            // Use index added by org.apache.pulsar.common.intercept.AppendIndexMetadataInterceptor if present.
+            // Requires exposingBrokerEntryMetadataToClientEnabled=true on brokers.
+            if (message.hasIndex()) {
+                final Optional<Long> index = message.getIndex();
+                if (index.isPresent()) {
+                    return index.get();
+                } else {
+                    LOGGER.debug("Found message {} with hasIndex=true but index is empty, using recordSequence",
+                            message.getMessageId());
+                }
+            } else {
+                LOGGER.debug("partitionerUseIndexAsOffset configured to true but no index found on the message {}, "
+                                + "perhaps the broker didn't exposed the metadata, using recordSequence",
+                        message.getMessageId());
+            }
```

Comment:
I don't think it's a good idea to implicitly fall back to the `recordSequence` if the user has configured the `partitionerUseIndexAsOffset` property. At the very least this should be a WARN log, or we should throw an exception here imo.

Reply:
I understand your concern. 
The problem is that who installs the sink may not know if the broker will put the index in the record metadata. 

In my first implementation this was a WARN level: https://github.com/streamnative/pulsar-io-cloud-storage/pull/470/commits/15659aafca1dd41182102a0a4fe11d6e16b60189#diff-ed19908cc0ff954ebf1f795a0e3fd708a3b1be501cda66c22ea1906d771da90cR91

This is the same behavior (except for the special batch messages handling) present in Kafka connect's connectors in Apache Pulsar: https://github.com/apache/pulsar/blob/4c22159f5a972e7a92382b9a90c6c70c43c5d166/pulsar-io/kafka-connect-adaptor/src/main/java/org/apache/pulsar/io/kafka/connect/KafkaConnectSink.java#L312-L359

To improve the usability we could:
- Explicitly state the behavior in the parameter doc
- Set those logs to WARN

Another (more complex) option would be to add another flag "noIndexAction" that regulates if throws error or accept records without index when `partitionerUseIndexAsOffset ` is true. 
Although, I would avoid to add another boilerplate option

Reply:
Yeah let's not add another flag :) I think moving the log to WARN and extending the parameter description is a good tradeoff